src/MultiFormats.Tests/MultiAddressTest.cs

```diff
@@ -0,0 +1,437 @@
+* (at your option) any later version.
+*
+* Catalyst.Node is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Catalyst.Node. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#endregion
+
+using System;
+using System.IO;
+using System.Net;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Newtonsoft.Json;
+
+namespace MultiFormats.Tests
+{
+    [TestClass]
+    public class MultiAddressTest
+    {
+        private const string Somewhere =
+            "/ip4/10.1.10.10/tcp/29087/ipfs/QmVcSqVEsvm5RR9mBLjwpb2XjFVn5bPdPL69mL8PH45pPC";
+
+        private const string Nowhere = "/ip4/10.1.10.11/tcp/29087/ipfs/QmVcSqVEsvm5RR9mBLjwpb2XjFVn5bPdPL69mL8PH45pPC";
+
+        [TestMethod]
+        public void Parsing()
+        {
+            var a = new MultiAddress(Somewhere);
+            Assert.AreEqual(3, a.Protocols.Count);
+            Assert.AreEqual("ip4", a.Protocols[0].Name);
+            Assert.AreEqual("10.1.10.10", a.Protocols[0].Value);
+            Assert.AreEqual("tcp", a.Protocols[1].Name);
+            Assert.AreEqual("29087", a.Protocols[1].Value);
+            Assert.AreEqual("ipfs", a.Protocols[2].Name);
+            Assert.AreEqual("QmVcSqVEsvm5RR9mBLjwpb2XjFVn5bPdPL69mL8PH45pPC", a.Protocols[2].Value);
+
+            Assert.AreEqual(0, new MultiAddress((string) null).Protocols.Count);
+            Assert.AreEqual(0, new MultiAddress("").Protocols.Count);
+            Assert.AreEqual(0, new MultiAddress("  ").Protocols.Count);
+        }
+
+        [TestMethod]
+        public void Unknown_Protocol_Name()
+        {
+            ExceptionAssert.Throws<FormatException>(() => new MultiAddress("/foobar/123"));
+        }
+
+        [TestMethod]
+        public void Missing_Protocol_Name() { ExceptionAssert.Throws<FormatException>(() => new MultiAddress("/")); }
+
+        [TestMethod]
+        public new void ToString() { Assert.AreEqual(Somewhere, new MultiAddress(Somewhere).ToString()); }
+
+        [TestMethod]
+        public void Value_Equality()
+        {
+            var a0 = new MultiAddress(Somewhere);
+            var a1 = new MultiAddress(Somewhere);
+            var b = new MultiAddress(Nowhere);
+            MultiAddress c = null;
+            MultiAddress d = null;
+
+            Assert.IsTrue(c == d);
+            Assert.IsFalse(c == b);
+            Assert.IsFalse(b == c);
+
+            Assert.IsFalse(c != d);
+            Assert.IsTrue(c != b);
+            Assert.IsTrue(b != c);
+
+#pragma warning disable 1718
+            Assert.IsTrue(a0 == a0);
+            Assert.IsTrue(a0 == a1);
+            Assert.IsFalse(a0 == b);
+
+#pragma warning disable 1718
+            Assert.IsFalse(a0 != a0);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Correct one of the identical expressions on both sides of operator '!='.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)